modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -1,143 +1,121 @@
+import eu.timepit.refined.types.numeric.NonNegLong
 import io.estatico.newtype.ops.toCoercibleIdOps
 import org.typelevel.log4cats.slf4j.Slf4jLogger
 
 object Rewards {
 
   def make[F[_]: Async: SecurityProvider](
-    config: RewardsConfig,
-    timeTriggerInterval: FiniteDuration,
-    softStaking: SoftStaking,
-    dtm: DTM,
-    stardust: StardustCollective
+    rewardsPerEpoch: SortedMap[EpochProgress, Amount],
+    softStaking: SimpleRewardsDistributor,
+    dtm: SimpleRewardsDistributor,
+    stardust: SimpleRewardsDistributor,
+    regular: RewardsDistributor[F]
   ): Rewards[F] =
     new Rewards[F] {
 
-      implicit val logger = Slf4jLogger.getLogger[F]
+      private val logger = Slf4jLogger.getLogger[F]
 
-      def calculateRewards(
+      def feeDistribution(
         epochProgress: EpochProgress,
+        transactions: SortedSet[transaction.Transaction],
         facilitators: NonEmptySet[Id]
       ): F[SortedSet[RewardTransaction]] = {
-        val rewards = getRewardsPerSnapshot(epochProgress)
-
-        if (rewards === Amount.empty) {
-          SortedSet.empty[RewardTransaction].pure[F]
-        } else
-          EitherT.liftF {
-            facilitators.toNonEmptyList
-              .traverse(_.toAddress)
-              .map(_.toNes)
-          }.flatMap(createSharedDistribution(epochProgress, rewards, _))
-            .flatMap { distribution =>
-              EitherT.fromEither {
-                Kleisli(weightByDTM)
-                  .compose(weightByStardust)
-                  .apply(distribution)
-              }
-            }
-            .flatMap { distribution =>
-              EitherT.fromEither {
-                distribution.toNel.traverse {
-                  case (address, reward) =>
-                    PosLong
-                      .from(reward.value)
-                      .map(reward => RewardTransaction(address, TransactionAmount(reward)))
-                      .left
-                      .map[RewardsError](NumberRefinementPredicatedFailure)
-                }.map(_.toNes.toSortedSet)
-              }
-            }
-            .flatTap(txs => EitherT.liftF(logger.info(s"Minted reward transactions: $txs")))
-            .valueOrF { error =>
-              val logError = error match {
-                case err @ NumberRefinementPredicatedFailure(_) =>
-                  logger.error(err)("Unhandled refinement predicate failure")
-                case IgnoredAllAddressesDuringWeighting =>
-                  logger.warn("Ignored all the addresses from distribution so weights won't be applied")
-              }
-
-              logError.as(SortedSet.empty[RewardTransaction])
-            }
+        val totalFee = transactions.toList
+          .map(_.fee)
+          .foldM(NonNegLong.MinValue) { case (acc, fee) => acc + fee.value }
+          .map(Amount(_))
+          .liftTo[F]
+
+        totalFee.flatMap { amount =>
+          regular.distribute(epochProgress, facilitators).run(amount).flatMap {
```

Comment:
I think it's better to base this distribution on `SnapshotOrdinal` instead of the `EpochProgress`. This would require passing a different param to `Random` so I think it's maybe better to pass the `Random` instance implicitly to the `Regular.distribute` method.

This would mean that we need to scrap the `RewardsDistributor` interface altogether since it constrains us to use two specific parameters `epochProgress: EpochProgress, facilitators: NonEmptySet[Id]`.

Reply:
Even if we use `SnapshotOrdinal` as a `seed` for both fee and time then making it implicit won't change much because `Random` is not a singleton, so I need to instantiate a new one for each snapshot anyway. Or maybe I didn't get it.

Scraping the `RewardDistributor` is fine because not all the programs use both `epochProgress` and `facilitators` so we end up with `case(_, _)` :+1:

Reply:
That's true, maybe it's better to pass it explicitly in this case.